docs/RFCS/certificate_rotation.md

```diff
@@ -0,0 +1,293 @@
+pushes them to the appropriate nodes.
+Keys are not kept by the trusted machine once deployed on the nodes.
+
+Upon renewal, filenames should differ allowing multiple sets of certificate/private key files
+to exist simultaneously.
+
+# Reloading certificate/key files
+
+The current method for providing certificates and private keys requires naming them specifically
+using flags or environment variables:
+* `--ca-cert`, `--ca-key` for CA certificate and key
+* `--cert`, `--key` for node or client certificate and key
+
+We can use a single file for CA certificates (see CA certificate renewal above) but we require
+multiple files for node and client certificates due to changing keys.
+
+We thus need to be able to specify a storage location containing an arbitrary number of
+certificates and keys.
+
+We propose a single directory containing all relevant certificates and private keys for
+the current node or client.
+
+## Directory
+
+The directory can be specified by command-line flag `--certs-dir` or environment variable
+`COCKROACH_CERTS_DIR`, with a default value of `cockroach-certs`.
+
+The permissions on the directory cannot be more permissive than `drwx------`.
+Directory ownership will not be checked (too much incompatibility across filesystems and
+architectures), relying instead on permissions to list the directory contents.
+
+## Files
+
+File requirements are as follows:
+* all files must be in the `COCKROACH_CERTS_DIR` directory, sub-directories are ignored
+* all files must be plain files, no sym-links allowed
+* maximum permissions for certificates: `-rwxr-xr-x`
+* maximum permissions for keys: `-rwx------`
+* all files are PEM-encoded
+
+File types and functions are determined automatically into categories by
+checking the following PEM or certificate information:
+
+| `pem.Block.Type` | `IsCA` | `ExtendedKeyUsage` | `Subject.CommonName` | Category |
+|---|---|---|---|---|
+| `CERTIFICATE` | `TRUE` | `CertSign | ContentCommitment` | Any | CA Certificate |
+| `CERTIFICATE` | `FALSE` | `ServerAuth | ClientAuth` | `Node` | Node certificate |
+| `CERTIFICATE` | `FALSE` | `ClientAuth` | not `Node` | Client certificate |
+| `PRIVATE KEY` | | | | Private Key |
+
+Any other combination (eg: `ExtendedKeyUsage = ServerAuth | ClientAuth` and `User = foo`) is invalid.
+
+Certificates and private keys can be matched by checking public-key components:
+* `rsa.PublicKey.N.Cmp(rsa.PrivateKey.N) == 0` for RSA keys.
+* `ecdsa.PublicKey.X.Cmp(ecdsa.PrivateKey.X) == 0 && ecdsa.PublicKey.Y.Cmp(ecdsa.PrivateKey.Y) == 0` for ecdsa keys.
+
+Details of PEM parsing and key checking can be found in [tls.X509KeyPair](https://golang.org/src/crypto/tls/tls.go?s=5915:5986#L183)
+
+CA certificates can be added to the certificate pool with little extra validation.
+
+Node certificates must be checked for validity. Specifically:
+* we must have a valid certificate/private key pair
+* the certificate must currently be valid (`Not Before < time.Now() < Not After`)
```

Comment:
I would recommend checking that they will be valid for some time in the future too, at least e.g. our reconnect timeouts (probably now + 1 day would be good for all intents and purposes)